apps/dashboard/src/components/primitives/field-editor/field-editor.tsx

```diff
@@ -0,0 +1,232 @@
+import { EditorView } from '@uiw/react-codemirror';
+import { Completion, CompletionContext } from '@codemirror/autocomplete';
+import { autocompletion } from '@codemirror/autocomplete';
+
+import { Editor } from '@/components/primitives/editor';
+import { completions } from '@/utils/liquid-autocomplete';
+import { LiquidVariable } from '@/utils/parseStepVariablesToLiquidVariables';
+import { Popover, PopoverTrigger } from '@/components/primitives/popover';
+import { createVariablePlugin } from './variable-plugin';
+import { VariablePopover } from './variable-popover';
+import { variablePillTheme } from './variable-plugin/variable-theme';
+import { useCallback, useMemo, useRef, useState, Dispatch, SetStateAction } from 'react';
+
+type SelectedVariable = {
+  value: string;
+  from: number;
+  to: number;
+} | null;
+
+type CompletionRange = {
+  from: number;
+  to: number;
+} | null;
+
+type FieldEditorProps = {
+  value: string;
+  onChange: (value: string) => void;
+  variables: LiquidVariable[];
+  placeholder?: string;
+  autoFocus?: boolean;
+  size?: 'default' | 'lg';
+  fontFamily?: 'inherit';
+  id?: string;
+};
+
+export function FieldEditor({
+  value,
+  onChange,
+  variables,
+  placeholder,
+  autoFocus,
+  size = 'default',
+  fontFamily = 'inherit',
+  id,
+}: FieldEditorProps) {
+  const viewRef = useRef<EditorView | null>(null);
+  const lastCompletionRef = useRef<CompletionRange>(null);
+
+  const { selectedVariable, setSelectedVariable, handleVariableSelect, isUpdatingRef } = useVariableSelection();
+
+  const handleVariableUpdate = useVariableUpdate(
+    selectedVariable,
+    viewRef,
+    isUpdatingRef,
+    onChange,
+    setSelectedVariable
+  );
+
+  const completionSource = useCompletionSource(variables, lastCompletionRef);
+
+  const extensions = useMemo(
+    () => [
+      autocompletion({
+        override: [completionSource],
+        closeOnBlur: false,
+        defaultKeymap: true,
+        activateOnTyping: true,
+      }),
+      EditorView.lineWrapping,
+      variablePillTheme,
+      createVariablePlugin({ viewRef, lastCompletionRef, onSelect: handleVariableSelect }),
+    ],
+    [variables, completionSource, handleVariableSelect]
+  );
+
+  return (
+    <div className="relative">
+      <Editor
+        size={size}
+        className="flex-1"
+        autoFocus={autoFocus}
+        fontFamily={fontFamily}
+        placeholder={placeholder}
+        id={id}
+        extensions={extensions}
+        value={value}
+        onChange={onChange}
+      />
+      <Popover
+        open={!!selectedVariable}
+        onOpenChange={(open) => {
+          if (!open) {
+            setTimeout(() => setSelectedVariable(null), 0);
+          }
+        }}
+      >
+        <PopoverTrigger asChild>
+          <div />
+        </PopoverTrigger>
+        {selectedVariable && (
+          <VariablePopover
```

Comment:
the popover jumps, I feel it should be aligned with the start of the variable, but not sure if that is possible

https://github.com/user-attachments/assets/07d3cc9f-83ca-41a8-a332-3a2ddf44ff9e